src/openzeppelin/tests/test_account.cairo

```diff
@@ -0,0 +1,330 @@
+        to: erc20.contract_address, selector: transfer_selector, calldata: calldata1
+    };
+
+    let mut calldata2 = ArrayTrait::new();
+    let amount2: u256 = 500;
+    calldata2.append(recipient2.into());
+    calldata2.append(amount2.low.into());
+    calldata2.append(amount2.high.into());
+    let call2 = Call {
+        to: erc20.contract_address, selector: transfer_selector, calldata: calldata2
+    };
+
+    calls.append(call1);
+    calls.append(call2);
+    let ret = account.__execute__(calls);
+
+    // Assert that the transfers were successful
+    assert(erc20.balance_of(account.contract_address) == 200, 'Should have remaining');
+    assert(erc20.balance_of(recipient1) == 300, 'Should have transferred');
+    assert(erc20.balance_of(recipient2) == 500, 'Should have transferred');
+
+    // Test return value
+    let mut call1_serialized_retval = ret.at(0).span();
+    let mut call2_serialized_retval = ret.at(1).span();
+    let call1_retval = Serde::<bool>::deserialize(ref call1_serialized_retval);
+    let call2_retval = Serde::<bool>::deserialize(ref call2_serialized_retval);
+    assert(call1_retval.unwrap(), 'Should have succeeded');
+    assert(call2_retval.unwrap(), 'Should have succeeded');
+}
+
+#[test]
+#[available_gas(2000000)]
+fn test_public_key_setter() {
+    let new_public_key = 0x4444;
+    testing::set_contract_address(ACCOUNT_ADDRESS());
+    testing::set_caller_address(ACCOUNT_ADDRESS());
+    Account::set_public_key(new_public_key);
+
+    let public_key = Account::get_public_key();
+    assert(public_key == new_public_key, 'Should update key');
+}
+
+#[test]
+#[available_gas(2000000)]
+#[should_panic(expected: ('Account: unauthorized', ))]
+fn test_public_key_setter_different_account() {
+    let new_public_key = 0x4444;
+    let caller = contract_address_const::<0x123>();
+    testing::set_contract_address(ACCOUNT_ADDRESS());
+    testing::set_caller_address(caller);
+    Account::set_public_key(new_public_key);
+}
+
+#[test]
+#[available_gas(2000000)]
+#[should_panic(expected: ('Account: invalid caller', ))]
+fn test_account_called_from_contract() {
+    let calls = ArrayTrait::new();
+    let caller = contract_address_const::<0x123>();
+    testing::set_contract_address(ACCOUNT_ADDRESS());
+    testing::set_caller_address(caller);
+    Account::__execute__(calls);
+}
+
+//
+// test internals
+//
+
+#[test]
+#[available_gas(2000000)]
+fn test__assert_only_self_true() {
+    testing::set_contract_address(ACCOUNT_ADDRESS());
+    testing::set_caller_address(ACCOUNT_ADDRESS());
+    Account::_assert_only_self();
+}
+
+#[test]
+#[available_gas(2000000)]
+#[should_panic(expected: ('Account: unauthorized', ))]
+fn test__assert_only_self_false() {
+    testing::set_contract_address(ACCOUNT_ADDRESS());
+    let other = contract_address_const::<0x4567>();
+    testing::set_caller_address(other);
+    Account::_assert_only_self();
+}
+
+#[test]
+#[available_gas(2000000)]
+fn test__is_valid_signature() {
```

Comment:
this should be checked as behavior on each function we want to test (validate_deploy, validate_declare, etc), so i'm not sure if we want to test functions that are not part of presets or modules' API.

otoh i'm not against anything being tested while it doesn't damage the tests design.

Reply:
just in case, what i meant by this is that we're doing the signature validation checks on this function instead of `__validate__`, `__validate_deploy__`, etc. now we're closer since we're testing invalid signatures, but we're not doing the same checks (e.g. invalid signature lengths).

it seems our testing suite relies on those functions to be using `_is_valid_signature` in them and that's not the right approach. tests suites should make no assumption about implementation details, but to assert a function's behavior.

these checks should be performed directly on any validation function when possible, including the public `is_valid_signature`. what i wasn't opposed to was to also test internal functions, although it's not necessary as long as they're not part of a module's API.

Reply:
Got it, makes sense now.